Abstract Texture

iwubcode opened the pull request:
This is a smaller PR of @phire 's #3880.  Specifically, it adopts his "AbstractTexture" for every backend.

This is simply a bit of code refactoring that will make things easier in the future.  There shouldn't be any functional changes.

lioncash commented on Source/Core/VideoBackends/OGL/OGLTexture.cpp:
```diff
@@ -0,0 +1,241 @@
+namespace OGL
+{
+
+static u32 s_Textures[8];
```
std::array

lioncash commented on Source/Core/VideoBackends/OGL/OGLTexture.cpp:
```diff
@@ -0,0 +1,241 @@
+#include "Common/GL/GLInterfaceBase.h"
```
Missing the copyright stuff.

lioncash replied:
`Common/CommonTypes.h`

iwubcode replied:
What's this for?  It doesn't include the necessary GLInterface needed to compile.

lioncash replied:
I meant to indicate that it's an include that should be included along with everything else, not to replace it (my bad).

iwubcode replied:
Ah ok, no problem.  Thanks for letting me know!

lioncash commented on Source/Core/VideoBackends/OGL/OGLTexture.h:
```diff
@@ -0,0 +1,37 @@
+{
+public:
+
+  GLuint texId;
```
These should be private and have default initializers, as well as the `m_` prefix.

lioncash commented on Source/Core/VideoBackends/OGL/OGLTexture.h:
```diff
@@ -0,0 +1,37 @@
+  GLuint texId;
+  GLuint framebuffer;
+
+  OGLTexture(const TextureConfig& config);
```
`explicit`

lioncash commented on Source/Core/VideoBackends/OGL/OGLTexture.h:
```diff
@@ -0,0 +1,37 @@
+namespace OGL
+{
+
+class OGLTexture : public AbstractTexture
```
This can be a final class.

lioncash commented on Source/Core/VideoBackends/OGL/OGLTexture.h:
```diff
@@ -0,0 +1,37 @@
+#include "Common/GL/GLUtil.h"
+
+#include "VideoCommon/AbstractTexture.h"
+#include "VideoCommon/BPStructs.h"
```
Why is this include necessary?

lioncash commented on Source/Core/VideoBackends/OGL/OGLTexture.h:
```diff
@@ -0,0 +1,37 @@
+#pragma once
```
Missing the copyright stuff.

lioncash commented on Source/Core/VideoBackends/OGL/OGLTexture.cpp:
```diff
@@ -0,0 +1,241 @@
+    glActiveTexture(GL_TEXTURE0 + s_ActiveTexture);
+}
+
+}
```
`}  // namespace OGL`

Missing a newline at the end of this file.

lioncash commented on Source/Core/VideoBackends/OGL/OGLTexture.h:
```diff
@@ -0,0 +1,37 @@
+  static void SetStage();
+};
+
+}
```
`}  // namespace OGL`

Also missing a newline at the end of this file.

lioncash commented on Source/Core/VideoBackends/OGL/OGLTexture.h:
```diff
@@ -0,0 +1,37 @@
+  virtual void FromRenderTarget(bool is_depth_copy, const EFBRectangle& srcRect, bool scaleByHalf,
+    unsigned int cbufid, const float* colmat) override;
+
+  static void DisableStage(unsigned int stage);
```
Why static behavior for these two in particular?

iwubcode replied:
These are called by outside classes (TextureConverter, Render, etc).  It will probably need to be cleaned up later.  A large majority of this code was following the design originally dictated by phire.  I modified it where I felt like I could.

lioncash commented on Source/Core/VideoBackends/OGL/TextureCache.h:
```diff
@@ -8,54 +8,45 @@
 namespace OGL
 {
+
+// TODO: Don't do this
```
I would really hope this would be resolved before trying to merge this. Otherwise the `s_` is wrong in this case, and this introduces a tremendous amount of global state.

iwubcode replied:
@phire and @stenzek and I talked about this, honestly there's not really a clean way to pass this around atm.  They have a grand vision..this is a simple stepping stone for them.

lioncash replied:
Sorry, but in that case, I absolutely refuse to agree to merging this if it introduces this much global state, even temporarily.

degasus replied:
A temporary fix would be to move them to OGLTexture.cpp as well, with an additional Init() and Shutdown() function.

lioncash commented on Source/Core/VideoBackends/Software/SWTexture.h:
```diff
@@ -0,0 +1,35 @@
+namespace SW
+{
+
+class SWTexture : public AbstractTexture
```
This can be a final class.

lioncash commented on Source/Core/VideoBackends/Software/SWTexture.h:
```diff
@@ -0,0 +1,35 @@
+class SWTexture : public AbstractTexture
+{
+public:
+
```
Unnecessary newline.

lioncash commented on Source/Core/VideoBackends/Software/SWTexture.h:
```diff
@@ -0,0 +1,35 @@
+#pragma once
```
Missing copyright stuff.

lioncash commented on Source/Core/VideoBackends/Software/SWTexture.h:
```diff
@@ -0,0 +1,35 @@
+{
+public:
+
+  SWTexture(const TextureConfig& config) : AbstractTexture(config)
```
`explicit`

Also all function implementations should be in a cpp file.

lioncash commented on Source/Core/VideoBackends/Vulkan/TextureCache.h:
```diff
@@ -54,34 +32,39 @@ class TextureCache : public TextureCacheBase
                           u32 aligned_height, u32 row_stride, const u8* palette,
                           TlutFormat palette_format) override;
 
+  VkShaderModule EFBColorToTexShader() const { return m_efb_color_to_tex_shader; }
```
Implementations should go in the cpp file.

lioncash commented on Source/Core/VideoBackends/Vulkan/VKTexture.cpp:
```diff
@@ -0,0 +1,327 @@
+  m_texture->TransitionToLayout(command_buffer, VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL);
+}
+
+}
```
`}  // namespace Vulkan`

also missing a newline at the end of the file.

lioncash commented on Source/Core/VideoBackends/Vulkan/VKTexture.h:
```diff
@@ -0,0 +1,41 @@
+
+class Texture2D;
+
+class VKTexture : public AbstractTexture
```
This can be a final class.

lioncash commented on Source/Core/VideoBackends/Vulkan/VKTexture.h:
```diff
@@ -0,0 +1,41 @@
+#include <vulkan/vulkan.h>
+
+#include "VideoCommon/AbstractTexture.h"
+#include "VideoCommon/BPStructs.h"
```
Why is this include needed?

lioncash commented on Source/Core/VideoBackends/Vulkan/VKTexture.h:
```diff
@@ -0,0 +1,41 @@
+{
+public:
+
+  Texture2D* GetRawTexIdentifier() const { return m_texture.get(); }
```
These two functions should have their implementations in the cpp file. They should also be placed after the virtual functions.

lioncash commented on Source/Core/VideoBackends/Vulkan/VKTexture.h:
```diff
@@ -0,0 +1,41 @@
+class VKTexture : public AbstractTexture
+{
+public:
+
```
Unnecessary newline.

lioncash commented on Source/Core/VideoBackends/Vulkan/VKTexture.h:
```diff
@@ -0,0 +1,41 @@
+  Texture2D* GetRawTexIdentifier() const { return m_texture.get(); }
+  VkFramebuffer GetFramebuffer() const { return m_framebuffer; }
+
+  VKTexture(const TextureConfig& config);
```
`explicit`

lioncash commented on Source/Core/VideoCommon/AbstractTexture.h:
```diff
@@ -0,0 +1,30 @@
+class AbstractTexture
+{
+public:
+  const TextureConfig config;
```
This should be private or protected. This should also come after functions.

lioncash commented on Source/Core/VideoCommon/AbstractTexture.h:
```diff
@@ -0,0 +1,30 @@
+public:
+  const TextureConfig config;
+
+  explicit AbstractTexture(const TextureConfig& c)
```
The definitions for the constructor and destructor should be in a cpp file. The destructor in particular can be defaulted in said cpp file.

lioncash commented on Source/Core/VideoCommon/AbstractTexture.h:
```diff
@@ -0,0 +1,30 @@
+    u32 level) = 0;
+  virtual void FromRenderTarget(bool is_depth_copy, const EFBRectangle& srcRect, bool scaleByHalf,
+    unsigned int cbufid, const float* colmat) = 0;
+};
```
Missing newline at the end of the file.

lioncash commented on Source/Core/VideoCommon/AbstractTexture.h:
```diff
@@ -0,0 +1,30 @@
+#pragma once
```
Missing copyright stuff.

lioncash commented on Source/Core/VideoCommon/TextureConfig.h:
```diff
@@ -0,0 +1,30 @@
+
+  bool operator==(const TextureConfig& o) const;
+
+  u32 width = 0;
```
These variable declarations should be placed after all functions.

lioncash commented on Source/Core/VideoCommon/TextureConfig.h:
```diff
@@ -0,0 +1,30 @@
+      return std::hash<u64>::operator()(id);
+    }
+  };
+};
```
Missing newline.

lioncash commented on Source/Core/VideoCommon/TextureConfig.h:
```diff
@@ -0,0 +1,30 @@
+#pragma once
```
Missing copyright stuff.

lioncash commented on Source/Core/VideoCommon/TextureConfig.cpp:
```diff
@@ -0,0 +1,14 @@
+MathUtil::Rectangle<int> TextureConfig::Rect() const
+{
+  return { 0, 0, static_cast<int>(width), static_cast<int>(height) };
+}
```
Missing newline.

lioncash commented on Source/Core/VideoCommon/TextureConfig.cpp:
```diff
@@ -0,0 +1,14 @@
+#include "VideoCommon/TextureConfig.h"
```
Missing copyright stuff.

lioncash commented on Source/Core/VideoCommon/TextureConfig.h:
```diff
@@ -0,0 +1,30 @@
+
+#include <functional>
+
+#include "Common/MathUtil.h"
```
`"Common/CommonTypes.h"` needs to be included.

lioncash commented on Source/Core/VideoCommon/TextureCacheBase.h:
```diff
@@ -194,38 +159,48 @@ class TextureCacheBase
-  typedef std::multimap<u32, TCacheEntryBase*> TexAddrCache;
-  typedef std::multimap<u64, TCacheEntryBase*> TexHashCache;
-  typedef std::unordered_multimap<TCacheEntryConfig, TCacheEntryBase*, TCacheEntryConfig::Hasher>
+
```
Unnecessary newline.

lioncash commented on Source/Core/VideoCommon/TextureCacheBase.h:
```diff
@@ -194,38 +159,48 @@ class TextureCacheBase
-  typedef std::multimap<u64, TCacheEntryBase*> TexHashCache;
-  typedef std::unordered_multimap<TCacheEntryConfig, TCacheEntryBase*, TCacheEntryConfig::Hasher>
+
+  struct TexPoolEntry  // Minimal version of TCacheEntry just for TexPool
```
Comments should go on top where possible.

lioncash commented on Source/Core/VideoCommon/TextureCacheBase.h:
```diff
@@ -102,44 +86,27 @@ class TextureCacheBase
 
     u64 CalculateHash() const;
+
+    u32 width() const { return texture->config.width; }
```
These should have their first letter uppercased.

lioncash commented on Source/Core/VideoCommon/TextureCacheBase.h:
```diff
@@ -61,23 +37,31 @@ class TextureCacheBase
-    std::unordered_set<TCacheEntryBase*> references;
+    std::unordered_set<TCacheEntry*> references;
+
+    TCacheEntry(std::unique_ptr<AbstractTexture> texture) : texture(std::move(texture)) {}
```
The constructor should be `explicit`.

The definition for the constructor and destructor should be within the cpp file.

lioncash commented on Source/Core/VideoBackends/Vulkan/VKTexture.h:
```diff
@@ -0,0 +1,41 @@
+#pragma once
```
Missing copyright stuff.

lioncash commented on Source/Core/VideoBackends/Vulkan/VKTexture.h:
```diff
@@ -0,0 +1,41 @@
+  VkFramebuffer m_framebuffer;
+};
+
+}
```
`}  // namespace Vulkan`

Missing a newline at the end of the file.

lioncash commented on Source/Core/VideoBackends/Software/SWTexture.h:
```diff
@@ -0,0 +1,35 @@
+  }
+};
+
+}
```
`}  // namespace SW`

lioncash commented on Source/Core/VideoBackends/Null/NullTexture.h:
```diff
@@ -0,0 +1,30 @@
+    unsigned int cbufid, const float* colmat) override {}
+};
+
+}
```
`}  // namespace Null`

lioncash commented on Source/Core/VideoBackends/D3D12/DX12Texture.h:
```diff
@@ -0,0 +1,40 @@
+  static void Shutdown();
+};
+
+}
```
`}  // namespace DX12`

lioncash commented on Source/Core/VideoBackends/D3D12/DX12Texture.h:
```diff
@@ -0,0 +1,40 @@
+class DX12Texture : public AbstractTexture
+{
+public:
+
```
Unnecessary newline

lioncash commented on Source/Core/VideoBackends/D3D12/DX12Texture.h:
```diff
@@ -0,0 +1,40 @@
+  D3D12_GPU_DESCRIPTOR_HANDLE m_texture_srv_gpu_handle = {};
+  D3D12_CPU_DESCRIPTOR_HANDLE m_texture_srv_gpu_handle_cpu_shadow = {};
+
+  DX12Texture(const TextureConfig& config);
```
`explicit`

lioncash commented on Source/Core/VideoBackends/D3D12/DX12Texture.cpp:
```diff
@@ -0,0 +1,314 @@
+  }
+}
+
+}
```
`}  // namespace DX12`

also missing a newline at the end of the file.

lioncash commented on Source/Core/VideoBackends/D3D/DXTexture.h:
```diff
@@ -0,0 +1,37 @@
+  D3DTexture2D* texture;
+};
+
+}
```
`}  // namespce DX11`

also missing a newline at the end of the file.

lioncash commented on Source/Core/VideoBackends/D3D/DXTexture.h:
```diff
@@ -0,0 +1,37 @@
+
+  D3DTexture2D* GetRawTexIdentifier() const { return texture; }
+
+  DXTexture(const TextureConfig& config);
```
`explicit`

lioncash commented on Source/Core/VideoBackends/D3D/DXTexture.h:
```diff
@@ -0,0 +1,37 @@
+{
+public:
+
+  D3DTexture2D* GetRawTexIdentifier() const { return texture; }
```
This should have its implementation in the cpp file. This should also be placed after the virtual functions.

lioncash commented on Source/Core/VideoBackends/D3D/DXTexture.cpp:
```diff
@@ -0,0 +1,237 @@
+    SAFE_RELEASE(efbcopycbuf[k]);
+}
+
+}
```
`}  // namespace DX11`

also missing a newline at the end of the file.

lioncash commented on Source/Core/VideoBackends/D3D/DXTexture.h:
```diff
@@ -0,0 +1,37 @@
+#pragma once
```
Missing copyright stuff.

lioncash commented on Source/Core/VideoBackends/D3D12/DX12Texture.cpp:
```diff
@@ -0,0 +1,314 @@
+#include "Common/Align.h"
```
Missing copyright stuff.

lioncash commented on Source/Core/VideoBackends/D3D/DXTexture.cpp:
```diff
@@ -0,0 +1,237 @@
+#include "Common/Logging/Log.h"
```
Missing copyright stuff.

degasus commented on Source/Core/VideoBackends/Null/NullTexture.h:
```diff
@@ -0,0 +1,31 @@
+  explicit NullTexture(const TextureConfig& config);
+  ~NullTexture();
+
+  virtual void Bind(unsigned int stage) override;
```
There is no need to mark all functions in the backends as virtual.

JosJuice replied:
Yes. The rule we use is that we don't write `virtual` if there also is an `override`.

degasus commented on Source/Core/VideoCommon/AbstractTexture.h:
```diff
@@ -0,0 +1,34 @@
+  virtual ~AbstractTexture()
+  {}
+
+  virtual void Bind(unsigned int stage) = 0;
```
I'm not sure if Bind() should be a function of the texture object. It is more a state of draw calls.

degasus commented on Source/Core/VideoCommon/AbstractTexture.h:
```diff
@@ -0,0 +1,34 @@
+  {}
+
+  virtual void Bind(unsigned int stage) = 0;
+  virtual bool Save(const std::string& filename, unsigned int level) = 0;
```
Feel free to provide a default implementation here. { return false; } is fine through, as we don't depend on this feature for emulation.

degasus commented on Source/Core/VideoCommon/AbstractTexture.h:
```diff
@@ -0,0 +1,34 @@
+  virtual void CopyRectangleFromTexture(const AbstractTexture* source,
+    const MathUtil::Rectangle<int>& srcrect,
+    const MathUtil::Rectangle<int>& dstrect) = 0;
+  virtual void Load(const u8* buffer, u32 width, u32 height, u32 expanded_width,
```
s/expanded_width/stride/

degasus commented on Source/Core/VideoCommon/AbstractTexture.h:
```diff
@@ -0,0 +1,34 @@
+    const MathUtil::Rectangle<int>& dstrect) = 0;
+  virtual void Load(const u8* buffer, u32 width, u32 height, u32 expanded_width,
+    u32 level) = 0;
+  virtual void FromRenderTarget(bool is_depth_copy, const EFBRectangle& srcRect, bool scaleByHalf,
```
This API is outdated since #4467

iwubcode replied:
Not sure what you mean?  I pulled this API directly from the class that already exists, this just moves it into a new location (and tidies up a few things)

lioncash commented on Source/Core/VideoBackends/D3D/DXTexture.cpp:
```diff
@@ -0,0 +1,243 @@
+// Copyright 2009 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
```
`<algorithm>`, `<cstddef>`, `"Common/CommonTypes.h"` need to be included.

iwubcode replied:
Do you want CommonTypes included even if it goes into the header?

lioncash replied:
Preferably, yes.

lioncash commented on Source/Core/VideoBackends/D3D/DXTexture.cpp:
```diff
@@ -0,0 +1,243 @@
+namespace DX11
+{
+const size_t MAX_COPY_BUFFERS = 32;
+ID3D11Buffer* efbcopycbuf[MAX_COPY_BUFFERS] = {0};
```
This can likely be static

iwubcode replied:
I prefer anonymous namespaces, any concerns?

lioncash replied:
either works, both the constant and the array can go in the anonymous namespace in that case.

lioncash commented on Source/Core/VideoBackends/D3D/DXTexture.h:
```diff
@@ -0,0 +1,39 @@
+
+#pragma once
+
+#include "VideoCommon/AbstractTexture.h"
```
`"Common/CommonTypes.h"` also needs to be included.

lioncash commented on Source/Core/VideoBackends/D3D12/DX12Texture.cpp:
```diff
@@ -0,0 +1,334 @@
+// Refer to the license.txt file included.
+
+#include "Common/Align.h"
+#include "Common/Assert.h"
```
`<algorithm>`, `<cstddef>`, `<cstring>` and `Common/CommonTypes.h` needs to be included as well.

lioncash commented on Source/Core/VideoBackends/D3D12/DX12Texture.h:
```diff
@@ -0,0 +1,47 @@
+
+#pragma once
+
+#include "VideoCommon/AbstractTexture.h"
```
`"Common/CommonTypes.h"` need to be included.

lioncash commented on Source/Core/VideoBackends/Null/NullTexture.cpp:
```diff
@@ -0,0 +1,36 @@
+{
+}
+
+NullTexture::~NullTexture()
```
This can be a defaulted destructor:

```cpp
NullTexture::~NullTexture() = default;
```

lioncash commented on Source/Core/VideoBackends/Null/NullTexture.h:
```diff
@@ -0,0 +1,30 @@
+
+#pragma once
+
+#include "VideoCommon/AbstractTexture.h"
```
`"Common/CommonTypes.h"` needs to be included.

lioncash commented on Source/Core/VideoBackends/OGL/OGLTexture.h:
```diff
@@ -0,0 +1,43 @@
+
+#pragma once
+
+#include "Common/GL/GLUtil.h"
```
`"Common/CommonTypes.h"` also needs to be included.

lioncash commented on Source/Core/VideoBackends/OGL/OGLTexture.h:
```diff
@@ -0,0 +1,43 @@
+class OGLTexture final : public AbstractTexture
+{
+public:
+  OGLTexture(const TextureConfig& tex_config);
```
This should be an `explicit` constructor.

lioncash commented on Source/Core/VideoBackends/Software/SWTexture.cpp:
```diff
@@ -0,0 +1,41 @@
+{
+}
+
+SWTexture::~SWTexture()
```
This can be a defaulted destructor:

```cpp
SWTexture::~SWTexture() = default;
```

lioncash commented on Source/Core/VideoBackends/Software/SWTexture.h:
```diff
@@ -0,0 +1,29 @@
+
+#pragma once
+
+#include "VideoCommon/AbstractTexture.h"
```
`"Common/CommonTypes.h"` needs to be included.

lioncash commented on Source/Core/VideoBackends/Vulkan/VKTexture.cpp:
```diff
@@ -0,0 +1,340 @@
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <stdexcept>
```
`<algorithm>`, `<cstddef>`, `<cstring>`, and `"Common/CommonTypes.h"` need to be included.

lioncash commented on Source/Core/VideoBackends/Vulkan/VKTexture.h:
```diff
@@ -0,0 +1,43 @@
+#include <memory>
+#include <vulkan/vulkan.h>
+
+#include "VideoCommon/AbstractTexture.h"
```
`"Common/CommonTypes.h"` also needs to be included.

lioncash commented on Source/Core/VideoCommon/AbstractTexture.h:
```diff
@@ -0,0 +1,30 @@
+class AbstractTexture
+{
+public:
+  explicit AbstractTexture(const TextureConfig& c) : config(c) {}
```
The definition for these functions should go in a cpp file.

lioncash commented on Source/Core/VideoCommon/TextureConfig.h:
```diff
@@ -0,0 +1,33 @@
+
+#pragma once
+
+#include <functional>
```
`<cstddef>` also needs to be included.

iwubcode commented:
@lioncash - I've removed the global state, please let me know if there are any other blockers
@degasus - I can't recall if it's you or @stenzek who does most of the OpenGL checkins.  Either way, if either of you could look at the opengl texture class (and texture cache) that'd be helpful.  I changed it to have member variables for the shader information (before they were file local statics).

I played a few more games, everything _seems_ to be working.  Let me know if you see differently.

nbohr1more commented:
Rebase please?

iwubcode commented:
@nbohr1more - done

iwubcode commented:
@stenzek - this now includes your "native compressed texture" support.  However I do not have any custom textures to test against.

I did do another around of general testing across various games (various backgrounds) not seeing any issues.

Fallcrest commented:
@BigheadSMZ Was doing testing of various cases for textures in that PR.

lioncash commented on Source/Core/VideoBackends/Vulkan/VKTexture.cpp:
```diff
@@ -0,0 +1,339 @@
+{
+}
+
+VKTexture* VKTexture::Create(const TextureConfig& tex_config)
```
This should return a `std::unique_ptr`.

iwubcode replied:
This gets passed to a  unique_ptr in Vulkan's TextureCache::CreateTexture (assuming it is non-null).  This is to get around the fact there's no return type on a constructor.  I wanted to use exceptions but eh.  The unique_ptr in TextureCache::CreateTexture is of type AbstractTexture.  I didn't see of a good way to "cast" the unique_ptr<VKTexture> to unique_ptr<AbstractTexture> so I opted for a raw ptr here instead.

Please let me know if you have any ideas.

iwubcode replied:
Andddd nm, looking into this more, passing the derived pointer might just work.  I assumed these special pointers were more restrictive.

lioncash replied:
unique_ptrs in a type hierarchy can implicitly convert to one another if they have the same deleter type (which is the case here). [Example](https://godbolt.org/g/CMpOlH)

iwubcode replied:
Thanks @lioncash, I literally just finished doing the same change as your example!  I'll go ahead and push those changes up.

iwubcode replied:
Seems like freebsd doesn't like that, gives me the same error that Intellisense gave me on Windows. 

`no viable conversion from returned value of type 'unique_ptr<Vulkan::VKTexture, default_delete<Vulkan::VKTexture>>' to function return type 'unique_ptr<AbstractTexture, default_delete<AbstractTexture>>'`

Regardless, I have a workaround.

lioncash commented:
Interpret that as an "approved once the above are addressed", since Github doesn't let you override a previous review with a more recent one without either explicit disapproval or approval :<

iwubcode commented on Source/Core/VideoCommon/AbstractTexture.cpp:
```diff
@@ -0,0 +1,41 @@
+{
+}
+
+AbstractTexture::~AbstractTexture()
```
You're saying AbsstractTexture::~AbstractTexture() = default can go in the cpp?  I never knew that!

iwubcode commented on Source/Core/VideoBackends/OGL/OGLTexture.cpp:
```diff
@@ -0,0 +1,332 @@
+    return GL_UNSIGNED_BYTE;
+  }
+}
+}
```
Thanks, I'll add this to the D3D one too

iwubcode commented:
All changes addressed.  It should be ready to merge assuming no one has any other concerns.  A big thank you to @lioncash for all his help!

Tilka commented on Source/Core/VideoBackends/Null/NullTexture.h:
```diff
@@ -0,0 +1,30 @@
+  {}
+  ~NullTexture() {}
+
+  virtual void Bind(unsigned int stage) override {}
```
Why?

lioncash commented:
The copyright date still isn't updated to 2017 in the new files added to the codebase (I believe I mentioned this to you on IRC if I'm not mistaken).

iwubcode commented:
@lioncash - done, sorry about that

iwubcode commented:
This has been updated to remove DX12.  The pr-android failure seems to be a common thing, some issue building the package.

Also, talking on IRC, @stenzek looked at the code and was happy with it.  :)

ligfx commented:
@dolphin-emu-bot rebuild

degasus commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -274,51 +252,29 @@ void TextureCacheBase::ScaleTextureCacheEntryTo(TextureCacheBase::TCacheEntryBas
-    *entry = newentry;
-    textures_by_address.emplace((*entry)->addr, *entry);
+    new_texture->CopyRectangleFromTexture(entry->texture.get(), srcrect, dstrect);
+    entry->texture.swap(new_texture);
```
Might you push new_texture to the pool again?

iwubcode replied:
I think, we don't need to do that because the entry is already in textures_by_address, it is just the entry's texture that is changing.

degasus replied:
By `auto new_texture = AllocateTexture(newconfig);`, you've allocate a new texture object from the pool. Here in swap, you've changed the texture of the texture cache entry. So the old texture object is now in `new_texture` and will be destroyed afterwards. We can just free it instead by pushing it to the pool:
`texture_pool.emplace(new_texture->config, TexPoolEntry(std::move(new_texture)));`

phire replied:
I think with the current code, this is indeed a rare event.

But I could see future hybrid xfb putting pressure on this code (and the current buggy version of the hybrid xfb pr actually shoves a texture through here every frame, for games with tiled xfb copies).

You could argue that because it's rare, there is a good chance the texture is of a weird size and  won't be reused, but probably idea to pool it anyway.

iwubcode replied:
@phire / @degasus - done.  I get it now, thanks for explaining it to me!

degasus commented on Source/Core/VideoCommon/TextureConfig.cpp:
```diff
@@ -0,0 +1,18 @@
+
+#include <tuple>
+
+bool TextureConfig::operator==(const TextureConfig& o) const
```
Did you see this function in profile? We have quite big maps based on this struct, so it might be better to move this implementation in the header.

iwubcode replied:
No, nothing like that.  I tend to prefer forward declaring things and reducing include files in headers.  That's why this was moved to the cpp.  I usually only put simple getters/setters in the header.

I didn't think of the performance implications of this, I wonder how heavy a change this is?

degasus commented on Source/Core/VideoBackends/OGL/TextureCache.h:
```diff
@@ -20,43 +24,26 @@ class TextureCache : public TextureCacheBase
-    void CopyRectangleFromTexture(const TCacheEntryBase* source,
-                                  const MathUtil::Rectangle<int>& srcrect,
-                                  const MathUtil::Rectangle<int>& dstrect) override;
+  SHADER GetColorCopyProgram() const;
```
What do you think about SHADER&? Right now, they are usually quite simple, but this may differ....

iwubcode replied:
I made this change, tested it and the pooling fix on a few games.  No issues.

degasus commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -274,51 +252,33 @@ void TextureCacheBase::ScaleTextureCacheEntryTo(TextureCacheBase::TCacheEntryBas
-    textures_by_address.emplace((*entry)->addr, *entry);
+	// At this point new_texture has the old texture in it,
+	// we can potentially reuse this, so let's move it back to the pool
+	texture_pool.emplace(new_texture->config, TexPoolEntry(std::move(new_texture)));
```
How did this change pass lint?

iwubcode replied:
Not sure!  Sorry about that, fixed.

stenzek commented on Source/Core/VideoCommon/AbstractTexture.h:
```diff
@@ -0,0 +1,32 @@
+#include "Common/MathUtil.h"
+#include "VideoCommon/TextureConfig.h"
+#include "VideoCommon/VideoCommon.h"
+
```
It's probably worth renaming HostTextureFormat from VideoCommon to AbstractTextureFormat, and moving it here.

iwubcode replied:
Would you mind if we saved this for another PR?  That would touch a bunch more files and I would like to avoid bloating this code review more than it is..

stenzek replied:
I doubt it's a large change, and the enumeration doesn't make sense being in VideoCommon.h.

iwubcode replied:
A simple search shows 14 files, some probably overlapping with other files in this PR but that's still 14 files for a change that really has little to do with this concept.  Unless you feel strongly about this, I'd rather save this cleanup for another PR.

stenzek replied:
Kinda. It should just be a matter of adding includes (and if those files are working with textures, they should include the texture header anyway).

iwubcode replied:
Alright, if that's how you feel, I will add this as well..

stenzek commented on Source/Core/VideoCommon/AbstractTexture.h:
```diff
@@ -0,0 +1,32 @@
+  virtual void FromRenderTarget(bool is_depth_copy, const EFBRectangle& src_rect,
+                                bool scale_by_half, unsigned int cbuf_id, const float* colmat) = 0;
+
+  bool IsCompressedHostTextureFormat(HostTextureFormat format);
```
This method should be static.

stenzek commented on Source/Core/VideoCommon/AbstractTexture.h:
```diff
@@ -0,0 +1,32 @@
+                                bool scale_by_half, unsigned int cbuf_id, const float* colmat) = 0;
+
+  bool IsCompressedHostTextureFormat(HostTextureFormat format);
+  size_t CalculateHostTextureLevelPitch(HostTextureFormat format, u32 row_length);
```
This method should be static.

stenzek commented on Source/Core/VideoCommon/AbstractTexture.h:
```diff
@@ -0,0 +1,32 @@
+  bool IsCompressedHostTextureFormat(HostTextureFormat format);
+  size_t CalculateHostTextureLevelPitch(HostTextureFormat format, u32 row_length);
+
+  const TextureConfig config;
```
Rename to m_config, to follow the Dolphin coding style.

iwubcode replied:
I did that, changed it to protected and introduced a helper.  Thanks.

stenzek commented on Source/Core/VideoCommon/TextureCacheBase.h:
```diff
@@ -103,53 +83,32 @@ class TextureCacheBase
 
     u64 CalculateHash() const;
+
+    u32 Width() const { return texture->config.width; }
```
Width() -> GetWidth()?
Height() -> GetHeight()?
Levels() -> GetLevels() or GetNumLevels()
Layers() -> GetLayers() or GetNumLayers()
Format() -> GetFormat()

stenzek commented on Source/Core/VideoBackends/Vulkan/TextureCache.h:
```diff
@@ -54,34 +31,40 @@ class TextureCache : public TextureCacheBase
                           u32 aligned_height, u32 row_stride, const u8* palette,
                           TlutFormat palette_format) override;
 
+  VkShaderModule EFBColorToTexShader() const;
```
EFBColorToTexShader() -> GetEFBColorToTexShader(), etc.

stenzek commented on Source/Core/VideoBackends/Vulkan/TextureCache.h:
```diff
@@ -54,34 +31,40 @@ class TextureCache : public TextureCacheBase
+  VkShaderModule EFBColorToTexShader() const;
+  VkShaderModule EFBDepthToTexShader() const;
+  VkRenderPass RenderPass() const;
+  StreamBuffer* TextureUploadBufferRaw() const;
```
This could probably just be GetTextureUploadBuffer()

iwubcode commented:
@stenzek - with the exception of what I commented on, everything else should be resolved.  I did not know what I was looking for in regard to the "validation" functionality.  Please take a look and let me know if anything seems odd.

stenzek commented on Source/Core/VideoCommon/AbstractTexture.h:
```diff
@@ -0,0 +1,35 @@
+                                        const MathUtil::Rectangle<int>& dstrect) = 0;
+  virtual void Load(u32 level, u32 width, u32 height, u32 row_length, const u8* buffer,
+                    size_t buffer_size) = 0;
+  virtual void FromRenderTarget(bool is_depth_copy, const EFBRectangle& src_rect,
```
This should be named "CopyEFBToTexture", or something similar, and moved to TextureCache, taking a TCacheEntry parameter.

iwubcode replied:
I'd be fine with "CopyRegionToTexture" or "CopyRegionToCacheEntry" because this is eventually going to include XFBs as well.  I'm fine with moving this for now, as being on AbstractTexture didn't really help or hurt us.

iwubcode replied:
Never mind, I see what you're saying now.  In Wii-terms, EFB is essentially a RenderTarget (I always forget that).. and even when we include XFBs those are just "textures".

stenzek commented on Source/Core/VideoBackends/Vulkan/VKTexture.cpp:
```diff
@@ -0,0 +1,348 @@
+                                         const MathUtil::Rectangle<int>& srcrect,
+                                         const MathUtil::Rectangle<int>& dstrect)
+{
+  TextureCache::GetInstance()->CopyRectangleFromTexture(
```
Move the logic from TextureCache to here.

stenzek commented on Source/Core/VideoBackends/Vulkan/TextureCache.h:
```diff
@@ -54,34 +31,40 @@ class TextureCache : public TextureCacheBase
                           u32 aligned_height, u32 row_stride, const u8* palette,
                           TlutFormat palette_format) override;
 
+  VkShaderModule GetEFBColorToTexShader() const;
```
These accessors can be dropped once FromRenderTarget is moved to TextureCache.

iwubcode replied:
FYI, it couldn't be dropped because now CopyTextureRectangle needs it.  Also the RenderPass is needed in two locations.

stenzek replied:
See my point below, I think the implementation is incorrect here. RenderPass should probably be renamed to "TextureCopyRenderPass", as there are many render passes in the backend. The raw suffix can be dropped from TextureUploadBuffer, since streambuffer is not abstracted in common.

stenzek commented on Source/Core/VideoBackends/Vulkan/TextureCache.h:
```diff
@@ -54,34 +31,40 @@ class TextureCache : public TextureCacheBase
 
   // Copies the contents of a texture using vkCmdCopyImage
-  void CopyTextureRectangle(TCacheEntry* dst_texture, const MathUtil::Rectangle<int>& dst_rect,
+  void CopyTextureRectangle(const VKTexture* dst_texture, const MathUtil::Rectangle<int>& dst_rect,
```
Move to VKTexture.

stenzek commented on Source/Core/VideoBackends/Vulkan/TextureCache.h:
```diff
@@ -54,34 +31,40 @@ class TextureCache : public TextureCacheBase
 
   // Copies (and optionally scales) the contents of a texture using a framgent shader.
-  void ScaleTextureRectangle(TCacheEntry* dst_texture, const MathUtil::Rectangle<int>& dst_rect,
+  void ScaleTextureRectangle(const VKTexture* dst_texture, const MathUtil::Rectangle<int>& dst_rect,
```
Move to VKTexture. Also, shouldn't be a const method (as it's modifying the texture).

stenzek commented on Source/Core/VideoBackends/Software/SWmain.cpp:
```diff
@@ -137,7 +114,6 @@ void VideoSoftware::InitBackendInfo()
   g_Config.backend_info.bSupportsComputeShaders = false;
   g_Config.backend_info.bSupportsInternalResolutionFrameDumps = false;
   g_Config.backend_info.bSupportsGPUTextureDecoding = false;
-  g_Config.backend_info.bSupportsST3CTextures = false;
```
Any reason why this statement is dropped?

iwubcode replied:
Good catch, this was on accident

stenzek commented on Source/Core/VideoBackends/OGL/TextureCache.h:
```diff
@@ -20,43 +24,26 @@ class TextureCache : public TextureCacheBase
-    void CopyRectangleFromTexture(const TCacheEntryBase* source,
-                                  const MathUtil::Rectangle<int>& srcrect,
-                                  const MathUtil::Rectangle<int>& dstrect) override;
+  const SHADER& GetColorCopyProgram() const;
```
These accessors can be dropped after moving the EFB copy logic to TextureCache.

iwubcode commented:
@stenzek - I'll fix lint bot tonight but otherwise all your points should be addressed.  Please let me know if you see anything else concerning!

stenzek commented on Source/Core/VideoBackends/Vulkan/VKTexture.cpp:
```diff
@@ -0,0 +1,365 @@
+    TextureCache::GetInstance()->GetRenderPass(),
+    g_object_cache->GetPassthroughVertexShader(),
+    g_object_cache->GetPassthroughGeometryShader(),
+    TextureCache::GetInstance()->GetEFBColorToTexShader());
```
Have you tested this? Scaling textures shouldn't use the EFB copy shader.

iwubcode replied:
I had to run out quickly this morning, so I only did light testing.  I did see that I picked the wrong shader here.  Thanks 👍

stenzek commented on Source/Core/VideoBackends/D3D/DXTexture.cpp:
```diff
@@ -0,0 +1,196 @@
+                                         const MathUtil::Rectangle<int>& srcrect,
+                                         const MathUtil::Rectangle<int>& dstrect)
+{
+  const auto srcentry = static_cast<const DXTexture*>(source);
```
I think the full type (const DXTexture*) would be clearer here, rather than auto.

stenzek commented on Source/Core/VideoBackends/Vulkan/TextureConverter.cpp:
```diff
@@ -174,8 +175,11 @@ void TextureConverter::ConvertTexture(TextureCache::TCacheEntry* dst_entry,
     int pad[2];
   };
 
+  auto source_texture = static_cast<VKTexture*>(src_entry->texture.get());
```
VKTexture* source_texture = ...

stenzek commented on Source/Core/VideoBackends/Vulkan/TextureConverter.cpp:
```diff
@@ -514,15 +519,16 @@ void TextureConverter::DecodeTexture(VkCommandBuffer command_buffer,
   dispatcher.Dispatch(groups.first, groups.second, 1);
 
   // Copy from temporary texture to final destination.
+  auto vulkan_tex_identifier = destination_texture->GetRawTexIdentifier();
```
Texture2D*

stenzek commented on Source/Core/VideoBackends/Vulkan/TextureConverter.cpp:
```diff
@@ -422,6 +426,7 @@ void TextureConverter::DecodeTexture(VkCommandBuffer command_buffer,
                                      u32 width, u32 height, u32 aligned_width, u32 aligned_height,
                                      u32 row_stride, const u8* palette, TlutFormat palette_format)
 {
+  auto destination_texture = static_cast<VKTexture*>(entry->texture.get());
```
VKTexture*

stenzek commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -1390,32 +1352,45 @@ void TextureCacheBase::CopyRenderTargetToTexture(u32 dstAddr, unsigned int dstFo
-TextureCacheBase::AllocateTexture(const TCacheEntryConfig& config)
+TextureCacheBase::TCacheEntry* TextureCacheBase::AllocateCacheEntry(const TextureConfig& config)
+{
+  auto texture = AllocateTexture(config);
```
std::unique_ptr&lt;AbstractTexture&gt; texture = ...

stenzek commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -1390,32 +1352,45 @@ void TextureCacheBase::CopyRenderTargetToTexture(u32 dstAddr, unsigned int dstFo
+  {
+    return nullptr;
+  }
+  auto cacheEntry = new TCacheEntry(std::move(texture));
```
TCacheEntry* cache_entry = ...

stenzek commented on Source/Core/VideoBackends/D3D/TextureCache.cpp:
```diff
@@ -356,14 +126,16 @@ void main(
+void TextureCache::ConvertTexture(TCacheEntry* destination, TCacheEntry* source, void* palette,
+                                  TlutFormat format)
 {
+  auto source_texture = static_cast<DXTexture*>(source->texture.get());
```
DXTexture* source_texture = ...

stenzek commented on Source/Core/VideoBackends/OGL/TextureCache.cpp:
```diff
@@ -654,43 +360,42 @@ bool TextureCache::CompileShaders()
 
-  TCacheEntry* entry = (TCacheEntry*)_entry;
-  TCacheEntry* unconverted = (TCacheEntry*)_unconverted;
+  auto source_texture = static_cast<OGLTexture*>(source->texture.get());
```
OGLTexture* source_texture = ...

stenzek commented on Source/Core/VideoBackends/Vulkan/TextureCache.cpp:
```diff
@@ -130,169 +142,39 @@ void TextureCache::CopyEFB(u8* dst, const EFBCopyFormat& format, u32 native_widt
-      config.width, config.height, config.levels, config.layers, vk_format, VK_SAMPLE_COUNT_1_BIT,
-      VK_IMAGE_VIEW_TYPE_2D_ARRAY, VK_IMAGE_TILING_OPTIMAL, usage);
-
+  auto texture = VKTexture::Create(config);
```
This could likely be just return VKTexture::Create(config);

stenzek commented on Source/Core/VideoBackends/Vulkan/TextureCache.cpp:
```diff
@@ -711,4 +335,77 @@ void TextureCache::DeleteShaders()
+                                       const EFBRectangle& src_rect, bool scale_by_half,
+                                       unsigned int cbuf_id, const float* colmat)
+{
+  auto texture = static_cast<VKTexture*>(entry->texture.get());
```
VKTexture* texture = ...

stenzek commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -274,51 +252,34 @@ void TextureCacheBase::ScaleTextureCacheEntryTo(TextureCacheBase::TCacheEntryBas
 
-  TCacheEntryBase* newentry = AllocateTexture(newconfig);
-  if (newentry)
+  auto new_texture = AllocateTexture(newconfig);
```
std::unique_ptr&lt;AbstractTexture&gt; new_texture = ...

stenzek commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -274,51 +252,34 @@ void TextureCacheBase::ScaleTextureCacheEntryTo(TextureCacheBase::TCacheEntryBas
-    {
-      newentry->textures_by_hash_iter = textures_by_hash.emplace((*entry)->hash, newentry);
-    }
+    auto srcrect = entry->texture->GetConfig().GetRect();
```
Are these locals needed anywhere else? Possibly move the GetRect() call to the CopyRectangleFromTexture line.

nbohr1more commented:
@dolphin-emu-bot rebuild

iwubcode commented:
I've tested this with my fifo comparison tool and there were no changes for any of the backends
